### PR TITLE
Scale Native Storage

### PR DESCRIPTION
**Pull request by kandeeban15**
Native Storage is scaled to handle simultaneous streaming graphs and each worker can accommodate more than one partition of a graph.

### CODE REVIEW

File: src/nativestore/RelationBlock.h [hunk 1]
```diff
@@ -103,12 +103,12 @@ class RelationBlock {
...
-    static const unsigned long BLOCK_SIZE;  // Size of a relation record block in bytes
-    static std::string DB_PATH;
-    static std::fstream *relationsDB;
-    static std::fstream *centralrelationsDB;
+    static thread_local unsigned int nextRelationIndex;
+    static thread_local unsigned int nextCentralRelationIndex;
+    static thread_local const unsigned long BLOCK_SIZE;  // Size of a relation record block in bytes
+    static thread_local std::string DB_PATH;
+    static thread_local std::fstream *relationsDB;
+    static thread_local std::fstream *centralrelationsDB;
```
Comment (miyurud) [inline src/nativestore/RelationBlock.h:111]:
```suggestion
    static thread_local std::fstream *centralRelationsDB;
```

File: src/server/JasmineGraphInstanceService.cpp [hunk 2]
```diff
@@ -4408,9 +4397,7 @@ static void check_file_accessible_command(int connFd, bool *loop_exit_p) {
     }
 }
 
-static void graph_stream_start_command(
-    int connFd, std::map<std::string, JasmineGraphIncrementalLocalStore *> &incrementalLocalStoreMap,
-    bool *loop_exit_p) {
+static void graph_stream_start_command(int connFd, InstanceStreamHandler& InstanceStreamHandler, bool *loop_exit_p) {
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:4400]:
```suggestion
static void graph_stream_start_command(int connFd, InstanceStreamHandler& instanceStreamHandler, bool *loop_exit_p) {
```

File: src/server/JasmineGraphInstanceService.cpp [hunk 3]
```diff
@@ -4445,20 +4432,7 @@ static void graph_stream_start_command(
...
-    JasmineGraphIncrementalLocalStore *incrementalLocalStoreInstance;
-
-    if (incrementalLocalStoreMap.find(graphIdentifier) == incrementalLocalStoreMap.end()) {
-        incrementalLocalStoreInstance =
-            JasmineGraphInstanceService::loadStreamingStore(graphId, partitionId, incrementalLocalStoreMap);
-    } else {
-        incrementalLocalStoreInstance = incrementalLocalStoreMap[graphIdentifier];
-    }
-    incrementalLocalStoreInstance->addEdgeFromString(nodeString);
+    InstanceStreamHandler.handleRequest(nodeString);
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:4435]:
```suggestion
    instanceStreamHandler.handleRequest(nodeString);
```

File: src/util/kafka/InstanceStreamHandler.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
+#include "InstanceStreamHandler.h"
```
Comment (miyurud) [inline src/util/kafka/InstanceStreamHandler.cpp:1]: Lets add standard license header here

File: src/util/kafka/InstanceStreamHandler.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+#include "../../localstore/incremental/JasmineGraphIncrementalLocalStore.h"
+#include "../Utils.h"
+#include "../logger/Logger.h"
+
+Logger InstanceStreamHandlerLogger;
```
Comment (miyurud) [inline src/util/kafka/InstanceStreamHandler.cpp:6]: Logger names usually defined in lower case. See for example,

`Logger data_publisher_logger;`

File: src/util/kafka/InstanceStreamHandler.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
+#ifndef INSTANCESTREAMHANDLER_H
```
Comment (miyurud) [inline src/util/kafka/InstanceStreamHandler.h:1]: Lets add standard license header here

File: src/util/kafka/StreamHandler.cpp [hunk 7]
```diff
@@ -87,19 +94,21 @@ void StreamHandler::listen_to_kafka_topic() {
         obj["source"] = sourceJson;
         obj["destination"] = destinationJson;
         obj["properties"] = edgeJson["properties"];
-        long temp_s = partitionedEdge[0].second;
-        long temp_d = partitionedEdge[1].second;
+        long part_s = partitionedEdge[0].second;
+        long part_d = partitionedEdge[1].second;
+        long temp_s = part_s % atoi((Utils::getJasmineGraphProperty("org.jasminegraph.server.nworkers")).c_str());
```
Comment (miyurud) [inline src/util/kafka/StreamHandler.cpp:99]: Shall we first assign `atoi((Utils::getJasmineGraphProperty("org.jasminegraph.server.nworkers")).c_str())` to a local variable and then use the local variable in the lines 99 and 100? That way we can avoid unnecessary cost of reading the config value.

**Review by miyurud — CHANGES_REQUESTED** [review]
I have put some comments. Please address them and reply to each of my comments.

File: src/nativestore/NodeManager.cpp [hunk 8]
```diff
@@ -25,13 +25,27 @@ limitations under the License.
 #include "PropertyLink.h"
 #include "RelationBlock.h"
 #include "iostream"
+#include <sys/stat.h>
 
 Logger node_manager_logger;
 pthread_mutex_t lockEdgeAdd;
 
+bool fileExists(const std::string& path) {
```
Comment (Ishad-M-I-M) [inline src/nativestore/NodeManager.cpp:33]: Function for check file existance already there in Utils: https://github.com/miyurud/jasminegraph/blob/master/src/util/Utils.h#L69.
Use it

File: src/nativestore/NodeManager.cpp [hunk 9]
```diff
@@ -25,13 +25,27 @@ limitations under the License.
...
+    std::ifstream file(path);
+    return file.is_open();
+}
+
+std::fstream* openFile(const std::string &path, std::ios_base::openmode mode) {
```
Comment (Ishad-M-I-M) [inline src/nativestore/NodeManager.cpp:38]: Its better to move this function to utils. Might be useful in other files as well

File: src/util/kafka/InstanceStreamHandler.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    std::string graphIdentifier = graphId + "_" + partitionId;
+    return graphIdentifier;
+}
+
+JasmineGraphIncrementalLocalStore *
+InstanceStreamHandler::loadStreamingStore(std::string graphId, std::string partitionId, map<std::string,
+                                          JasmineGraphIncrementalLocalStore *> &graphDBMapStreamingStores) {
+    std::string graphIdentifier = graphId + "_" + partitionId;
+    InstanceStreamHandlerLogger.log("###INSTANCE### Loading streaming Store for" + graphIdentifier
+                                    + " : Started", "info");
```
Comment (Ishad-M-I-M) [inline src/util/kafka/InstanceStreamHandler.cpp:84]:
```suggestion
    InstanceStreamHandlerLogger.info("###INSTANCE### Loading streaming Store for" + graphIdentifier
                                    + " : Started");
```

File: src/util/kafka/InstanceStreamHandler.cpp [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    std::string folderLocation = Utils::getJasmineGraphProperty("org.jasminegraph.server.instance.datafolder");
+    auto *jasmineGraphStreamingLocalStore = new JasmineGraphIncrementalLocalStore(
+                                     stoi(graphId), stoi(partitionId));
+    graphDBMapStreamingStores.insert(std::make_pair(graphIdentifier, jasmineGraphStreamingLocalStore));
+    InstanceStreamHandlerLogger.log("###INSTANCE### Loading Local Store : Completed", "info");
```
Comment (Ishad-M-I-M) [inline src/util/kafka/InstanceStreamHandler.cpp:89]:
```suggestion
    InstanceStreamHandlerLogger.info("###INSTANCE### Loading Local Store : Completed");
```

File: src/nativestore/NodeManager.cpp [hunk 12]
```diff
@@ -264,9 +305,12 @@ void NodeManager::addNodeIndex(std::string nodeId, unsigned int nodeIndex) {
         char nodeIDC[NodeManager::INDEX_KEY_SIZE] = {0};  // Initialize with null chars
         std::strcpy(nodeIDC, nodeId.c_str());
         index_db.write(nodeIDC, sizeof(nodeIDC));
-        index_db.write(reinterpret_cast<char *>(&(nodeIndex)), sizeof(unsigned int));
-        node_manager_logger.debug("Writing node index --> Node key = " + std::string(nodeIDC) + " value " +
-                                  std::to_string(nodeIndex));
+        index_db.write(reinterpret_cast<char *>(&nodeIndex), sizeof(unsigned int));
+        index_db.flush();
+        node_manager_logger.debug("Writing node index --> Node key = " +
```
Comment (Ishad-M-I-M) [inline src/nativestore/NodeManager.cpp:310]:
```suggestion
        node_manager_logger.info("Writing node index --> Node key = " +
```
Debug logs only appear if debug mode is enabled. I thinks its better to be `info` here

File: src/nativestore/NodeManager.cpp (hunk 8 above)
Comment (kandeeban15) [reply src/nativestore/NodeManager.cpp:33]: But it doesn't check whether we have read permission for the file.

File: src/nativestore/NodeManager.cpp (hunk 12 above)
Comment (kandeeban15) [reply src/nativestore/NodeManager.cpp:310]: They should not be logged in production mode.

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 11 above)
Comment (kandeeban15) [reply src/util/kafka/InstanceStreamHandler.cpp:89]: Done.

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 10 above)
Comment (kandeeban15) [reply src/util/kafka/InstanceStreamHandler.cpp:84]: Done.

File: src/nativestore/RelationBlock.h (hunk 1 above)
Comment (kandeeban15) [reply src/nativestore/RelationBlock.h]: Done.

File: src/server/JasmineGraphInstanceService.cpp (hunk 3 above)
Comment (kandeeban15) [reply src/server/JasmineGraphInstanceService.cpp]: Done.

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 4 above)
Comment (kandeeban15) [reply src/util/kafka/InstanceStreamHandler.cpp:14]: Done.

File: src/util/kafka/InstanceStreamHandler.h (hunk 6 above)
Comment (kandeeban15) [reply src/util/kafka/InstanceStreamHandler.h:14]: Done.

File: src/util/kafka/StreamHandler.cpp (hunk 7 above)
Comment (kandeeban15) [reply src/util/kafka/StreamHandler.cpp]: Done.

File: src/nativestore/NodeManager.cpp (hunk 9 above)
Comment (kandeeban15) [reply src/nativestore/NodeManager.cpp]: Done.

File: src/nativestore/NodeManager.cpp (hunk 8 above)
Comment (Ishad-M-I-M) [reply src/nativestore/NodeManager.cpp]: But this is not a good method of checking it. 
You can check it by 
```cpp
access(fileName.c_str(), R_OK) == 0
```

Comment (Ishad-M-I-M) [reply src/nativestore/NodeManager.cpp]: I think checking for read permission is not necessary as jasminegraph is anyway running in `root` mode

File: src/nativestore/RelationBlock.h (hunk 1 above)
Comment (miyurud) [reply src/nativestore/RelationBlock.h]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 2 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 3 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 5 above)
Comment (miyurud) [reply src/util/kafka/InstanceStreamHandler.cpp]: Ok

File: src/util/kafka/InstanceStreamHandler.h (hunk 6 above)
Comment (miyurud) [reply src/util/kafka/InstanceStreamHandler.h:14]: Ok

File: src/util/kafka/StreamHandler.cpp (hunk 7 above)
Comment (miyurud) [reply src/util/kafka/StreamHandler.cpp]: Ok

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 4 above)
Comment (miyurud) [reply src/util/kafka/InstanceStreamHandler.cpp:14]: Ok

File: src/nativestore/NodeManager.cpp (hunk 12 above)
Comment (miyurud) [reply src/nativestore/NodeManager.cpp:295]: I also believe we should keep this log at debug level. Info level logging is unnecessarily verbose.

File: src/util/kafka/InstanceStreamHandler.cpp (hunk 11 above)
Comment (miyurud) [reply src/util/kafka/InstanceStreamHandler.cpp]: Ok

File: src/nativestore/NodeManager.cpp (hunk 9 above)
Comment (miyurud) [reply src/nativestore/NodeManager.cpp]: Ok

File: src/nativestore/NodeManager.cpp (hunk 8 above)
Comment (miyurud) [reply src/nativestore/NodeManager.cpp]: @kandeeban15 Please let us know how you will resolve this.

Comment (miyurud) [reply src/nativestore/NodeManager.cpp]: > I think checking for read permission is not necessary as jasminegraph is anyway running in `root` mode
@Ishad-M-I-M 
Why do we need to run JasmineGraph in root mode? So far we did not require it to run on root mode.

Comment (thevindu-w) [reply src/nativestore/NodeManager.cpp]: It's running in docker, so it will have root privileges in container by default

Comment (miyurud) [reply src/nativestore/NodeManager.cpp]: We will check for read permissions using a utility function defined in the Utils.cpp.

**Review by miyurud — APPROVED** [review]

File: src/util/Utils.cpp [hunk 13]
```diff
@@ -666,3 +666,18 @@ int Utils::createDatabaseFromDDL(const char *dbLocation, const char *ddlFileLoca
     util_logger.info("Database created successfully");
     return 0;
 }
+
+bool Utils::fileExistsWithReadPermission(const string &path) {
+    std::ifstream file(path);
+    return file.is_open();
```
Comment (thevindu-w) [inline src/util/Utils.cpp:672]:
```suggestion
    return access(fileName.c_str(), R_OK) == 0;
```

Comment (thevindu-w) [inline src/util/Utils.cpp:672]:
```suggestion
    return access(fileName.c_str(), R_OK) == 0;
```

We do not need to open the file to check for permissions

Comment (kandeeban15) [reply src/util/Utils.cpp]: Done.

Comment (miyurud) [reply src/util/Utils.cpp]: Ok

**Pull request merged**